### DISCUSSION TITLE
Smart UTXO (hooks) - AEIP5

### DISCUSSION THREAD

**Discussion by samuelmanzanera** [archethic-foundation/aeip#23] [AEIP]
This is discussion for the AEIP5 located at: https://github.com/archethic-foundation/aeip/blob/main/AEIP-05.md

**Comment by Neylix** [comment]
Nice work !
Some remarks: 
- In the language support you say the interpreter cannot directly be used, but we can create 2 new block actions and conditions like `condition hooks: []` and `actions triggered_by: hooks do ... end` these conditions and actions would be allowed only in a hooks and not in a code part of a transaction
- Is the hooks attached to the token or the utxo ? I mean a hooks can be defined at the token creation and follow any utxo of this token even if we split the utxo in multiple ones, or the hooks is just for one utxo and then it will be erased on the next transfer ?

**Comment by samuelmanzanera** [comment]
The initial idea was to specify the hooks during the transfer, so they would be propagated through the UTXOs then.

Moreover, the thing with the hooks was to avoid calling smart contract each time, but executing some predetermined code.
So, I don't see well the usage of the `condition hooks: []` or `actions triggered_by: hooks do ... end`

For the token, we could inject hook when the UTXO are created.
But if there is not pre-hook, someone could add some in the transfers.

We just have to see how to prevent overwriting existing hook or a merge strategy.
